### DISCUSSION TITLE
`git-revision-date-localized-plugin` Error while running `make serve`

### DISCUSSION THREAD

**Discussion by tomas-svoboda-ipf** [squidfunk/mkdocs-material#7611] [Q&A]
Hello community,

I would like to ask you for a hint how to solve following bug I am getting while trying to run make serve.

`ERROR   -  [git-revision-date-localized-plugin] Unable to read git logs of '/docs/docs'.  To ignore this error, set option 'fallback_to_build_date: true'`

I read 2 different Discussions and honestly, none of them is relevant as they are over a year old and using older registries (before you ask, YES I did try to do suggested workarounds from other Threads/Discussions).

I am using following registry: `9.5.40-insiders-4.53.14`
My git version: `2.39.3 (Apple Git-146)`

Yes, I tried to put `fallback_to_build_date: true` into `mkdocs.yml` and removed when Error was still here.

My entire docker command under make serve: `docker run -it --rm -u 501:20 --name mkdocs -p 8000:8000 -v /Users/user/git/docs:/docs registry.gitlab.com/company/documentation/docs:9.5.40-insiders-4.53.14`
By the way, this resulted into another Warning message: `The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested but I solved this by docker run --platform linux/amd64 -it --rm -u 501:20 --name mkdocs -p 8000:8000 -v /Users/user/git/docs:/docs registry.gitlab.com/company/documentation/docs:9.5.40-insiders-4.53.14.`

Anyway, `ERROR   -  [git-revision-date-localized-plugin] Unable to read git logs of '/docs/docs'.  To ignore this error, set option 'fallback_to_build_date: true'` remains.

Any advised more than appreciated.

**Comment by squidfunk** [comment]
So it looks like you're building a custom Docker image, and running into troubles. Note that we currently do not provide 1:1 support on this forums for customization, and a custom Docker image is definitely considered a customization. The reason is that with 1:1 support, we're investing our time in helping one user, with the majority of users not benefiting. This time must then be taken out of the time we invest to fix bugs or develop new features. If you're interested in 1:1 support, let us know at sponsors@squidfunk.com, as we're currently evaluating ways to make that happen ☺️

Other than that, maybe a community member can help out here.

**Comment by tomas-svoboda-ipf** [comment]
`If you're interested in 1:1 support, let us know at "sponsors@squidfunk.com"`. A private session? That looks so expensive. 😄 
Honestly, I thought this seems to be quite common problem as same error message occurred already few times on this project. And always, it was some known problem fixed by small adjust in `mkdoc` file. So, I hoped into same scenario.
I will try to figure it out myself first, if I am not able to fix this myself for days/weeks, I will ask you again through email.

**Reply by squidfunk** [reply]
I was not talking about a private session, which is of course possible. We're currently evaluating options how we can better offer support for enterprises, given our time budget and the available funding we have.

**Comment by kamilkrzyskow** [comment]
Hello @tomas-solka-ipf,
I don't use Docker so please bare with me 😅 I don't have a solution per se, but perhaps some questions that could point into the right direction.

First about what you said:
> I read 2 different Discussions and honestly, none of them is relevant as they are over a year old and using older registries (before you ask, YES I did try to do suggested workarounds from other Threads/Discussions).

What are those 2 different Discussions, instead of making general comments better to provide the source information itself.
And you first posted this thread in MkDocs Discussions, and were pointed here, this also would be worth mentioning 🫡 
- https://github.com/mkdocs/mkdocs/discussions/3848

> [...] as same error message occurred already few times on this project. And always, it was some known problem fixed by small adjust in mkdoc file

Again, no sources were given, so I'm not seeing what you're seeing. If the project already had this issue with the `git-revision-date-localized-plugin` error, then the solution should be documented with the cause?

---

As for the Docker part to add to the previous answer in the other discussion:
> git-revision-date-localized-plugin needs the Git history

> `Unable to read git logs of '/docs/docs'`

So my question is what does the mounted docs directory contain? Is this your repository root with `mkdocs.yml` and `.git` directory?
`-v /Users/user/git/docs:/docs`
Or is the `mkdocs.yml` contained in the Docker image?
`registry.gitlab.com/company/documentation/docs:9.5.40-insiders-4.53.14`

Looking at the docs:
- https://docs.docker.com/reference/cli/docker/container/run/#volume
`docker run -v $(pwd):$(pwd) -w $(pwd)`

It does look like there are separate flags for setting the mount `-v` and working directory `-w`
And the `git-revision-date-localized-plugin` plugin looks for the Git repository in the parent directories of the `docs_dir`, so it seems like it's not being included together with the `/docs` files, or at least it's not in the parent directory of the `docs_dir` processed with MkDocs 🤔

**Reply by tomas-svoboda-ipf** [reply]
>What are those 2 different Discussions, instead of making general comments better to provide the source information itself.

[https://github.com/squidfunk/mkdocs-material/issues/5336](url) and [https://github.com/mkdocs/mkdocs/discussions/2843](url)

>Again, no sources were given, so I'm not seeing what you're seeing. If the project already had this issue with the git-revision-date-localized-plugin error, then the solution should be documented with the cause?

By this, I mean here on github. When I was searching for the problem I got, I was forwarded to 2 threads mentioned above.
///
I was looking to the output message for `make serve` again few minutes back and I totally missed (rookie mistake), probably the most important section of entire output: 
```
git.exc.GitCommandError: Cmd('git') failed due to: exit code(128)
  cmdline: git log --date=unix --format=%at -n 1 --no-show-signature /docs/docs
  stderr: 'fatal: detected dubious ownership in repository at '/docs'
To add an exception for this directory, call:

	git config --global --add safe.directory /docs'
make: *** [serve] Error 1
```
I tried to run `git config --global --add safe.directory /docs` command but problem is still here. 

```
git config --global --get-all safe.directory

/docs
```
Above, you can see what are allowed directories, but as @kamilkrzyskow said: `please bare with me`. So, I am rookie in entire GIT, so please bear with me 😄.

**Reply by kamilkrzyskow** [reply]
Thanks for the follow up. The linked discussion:
- https://github.com/mkdocs/mkdocs/discussions/2843

Is exactly related to the same thing you're facing with the `safe.directory`, and disproves the previous statement about it being always resolved with some `mkdocs.yml` file change ^^

As for the answer in that other discussion thread, it was answered on the 16th of May 2022 and fortunately for us [NicolasMassart](https://github.com/NicolasMassart) has a public profile.

So checking the dates between the 13th and 16th I found his PR:
- https://github.com/NicolasMassart?tab=overview&from=2022-05-13&to=2022-05-16
- https://github.com/Consensys/doctools.action-builder/pull/14

Looking at the changes in the PR it was fixed by modifying the [Docker image itself](https://github.com/Consensys/doctools.action-builder/pull/14/files#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557R21)
So perhaps you need to modify the `registry.gitlab.com/company/documentation/docs` image.

> I tried to run git config --global --add safe.directory /docs command but problem is still here.

Did you try to run it inside the Docker image? I see you're running the `-it` flags in your command so there should be some sort of interactive terminal? However, I guess the Docker image runs `mkdocs serve` first with the error and only afterwards you can stop MkDocs and run the `git` command. Or perhaps when MkDocs terminates it also terminates the whole container? 🤔 

I believe you can't run the `git` command outside of Docker even with the `--global` flag, and expect it to work inside Docker because it creates an isolated container 🤔 

I don't know if this can influence the result, but you're using the `-u 501:20` flag to set the UID:GID so perhaps changing this could "trick" the image into believing it's the owner of the files if the UID/GID match, but that's likely too easy 😆 Still, inside the Docker container run `git -lan` to see the UID/GID and try to match them in the `-u` flag?

**Reply by tomas-svoboda-ipf** [reply]
Well, [https://github.com/mkdocs/mkdocs/discussions/2843](url) is 2 years old, advice here is: `The most effective way to protect against this vulnerability is to upgrade to Git v2.35.2.` but I am on v2.39.3.

Maybe I am just too naive but I expect the problem shouldn't reoccur on newer version.
So, `Looking at the changes in the PR it was fixed by modifying the Docker image itself` did not help. 

I do not understand to `modify  registry.gitlab.com/company/documentation/docs` I have a feeling that this is not something what should be editable on CLI, pasting this into the browser gave me `404 - page not found`, so I am just confused.

To be fair, I have almost no idea what you are talking about for last 3 sections 😄 . Give me please some time to understand to your words 😄 .

Anyway, what I am supposed to used instead of `-u 501:20`?

**Reply by kamilkrzyskow** [reply]
> but I expect the problem shouldn't reoccur on newer version

This train of thought can apply if the new version fixes some bug, but even then bugs can resurface later.
The issue concerns a security vulnerability, and the fix in `2.35.2` is exactly this behaviour where it prevents the loading of the directory if not a matching user was detected, and this new behaviour is here to stay even in `2.39.3`

> The most effective way to protect against this vulnerability is to upgrade to [Git v2.35.2](https://github.blog/open-source/git/git-security-vulnerability-announced/#). **This version changes Git’s behavior when looking for a top-level .git directory to stop when its directory traversal changes ownership from the current user.** (If you wish to make an exception to this behavior, you can use the new multi-valued [safe.directory](https://git-scm.com/docs/git-config/2.35.2#Documentation/git-config.txt-safedirectory) configuration).

---

> I do not understand to modify  registry.gitlab.com/company/documentation/docs [...] 404 - page not found, so I am just confused.

I'm the one confused here 😅 

`registry.gitlab.com/company/documentation/docs` is the URL you've provided in the OP:
- `registry.gitlab.com/company/documentation/docs:9.5.40-insiders-4.53.14`
  I understand the `/company/` part is obfuscated, but it's up to you to fill in the blanks

This is your Docker image, which executes some commands from the Dockerfile like here:
- https://hub.docker.com/layers/squidfunk/mkdocs-material/latest/images/sha256-0e60104c5596b1e1584b2b113d8ddacd10000dfbe36e3fc4be5b93429c0eac5b?context=explore

I get that the Image was modified by your company to include the Insiders version of Material for MkDocs, so best to contact the person responsible for that to consider adding those changes from:
- https://github.com/Consensys/doctools.action-builder/pull/14

or try `git config --global --add safe.directory '*'` to allow every directory and not just `docs/`

---

> what I am supposed to used instead of -u 501:20?

I don't know, I can assume the default Linux UID/GID is 1000 so you could try `-u 1000:1000`, Like I said this hack seems to be too easy to work.

```
$ ls -lan
total 2062737
drwxr-xr-x 1 196509 196921          0 Oct 14 06:23  ./
drwxr-xr-x 1 196509 196921          0 Apr  2  2021  ../
-rw-r--r-- 1 196509 196921        555 Jun 19 15:48  r.txt
```

In my case on Bash for Windows it shows UID:196509 and GID:196921. This is just an example how to read the results from `ls -lan`.
